Video trimming plugin

DanielHuntleySBG opened the pull request:
### Link to related issue (if applicable)

### Summary of proposed changes

### Checklist
- [x] Use `develop` as the base branch
- [ ] Exclude the gulp build (`/dist` changes) from the PR
- [ ] Test on [supported browsers](https://github.com/sampotts/plyr#browser-support)

sampotts committed:
WIP

sampotts committed:
Merge branch 'develop' into css-variables

# Conflicts:
#	demo/dist/demo.css
#	demo/dist/demo.js
#	demo/dist/demo.js.map
#	demo/dist/demo.min.js
#	demo/dist/demo.min.js.map
#	dist/plyr.css
#	dist/plyr.js
#	dist/plyr.js.map
#	dist/plyr.min.js
#	dist/plyr.min.js.map
#	dist/plyr.polyfilled.js
#	dist/plyr.polyfilled.js.map
#	dist/plyr.polyfilled.min.js
#	dist/plyr.polyfilled.min.js.map
#	gulpfile.js
#	src/sass/components/captions.scss
#	src/sass/components/control.scss

sampotts committed:
Work on outline/focus styles

sampotts committed:
Merge branch 'develop' into css-variables

# Conflicts:
#	demo/dist/demo.css
#	demo/index.html
#	dist/plyr.css
#	gulpfile.js
#	package.json
#	yarn.lock

sampotts committed:
Fix merge

sampotts committed:
More work on variable usage

sampotts committed:
Merge branch 'develop' into css-variables

# Conflicts:
#	.eslintrc
#	demo/dist/demo.css
#	demo/dist/demo.js
#	demo/dist/demo.min.js
#	demo/dist/demo.min.js.map
#	dist/plyr.css
#	dist/plyr.js
#	dist/plyr.min.js
#	dist/plyr.min.js.map
#	dist/plyr.min.mjs
#	dist/plyr.min.mjs.map
#	dist/plyr.mjs
#	dist/plyr.polyfilled.js
#	dist/plyr.polyfilled.min.js
#	dist/plyr.polyfilled.min.js.map
#	dist/plyr.polyfilled.min.mjs
#	dist/plyr.polyfilled.min.mjs.map
#	dist/plyr.polyfilled.mjs
#	gulpfile.js
#	package.json

sampotts committed:
Clean up

sampotts committed:
Fixed merge

sampotts committed:
Build

nisarhassan12 committed:
update the gitpod setup description to be more precise.

sampotts committed:
Merge pull request #1731 from nisarhassan12/master

update the gitpod setup description to be more precise.

sampotts committed:
Merge branch 'develop' into css-variables

# Conflicts:
#	demo/dist/demo.css
#	demo/dist/demo.min.js.map
#	demo/index.html
#	dist/plyr.css
#	dist/plyr.min.js.map
#	dist/plyr.min.mjs.map
#	dist/plyr.polyfilled.min.js.map
#	dist/plyr.polyfilled.min.mjs.map
#	gulpfile.js
#	src/sass/base.scss
#	src/sass/components/control.scss
#	src/sass/settings/colors.scss
#	src/sass/settings/controls.scss

sampotts committed:
Formatting fixes

sampotts committed:
Sentry upgrade for demo

Danielh112 committed:
Structural code for video trimming UI

Danielh112 committed:
Merge remote-tracking branch 'origin/develop' into video-trimming-plugin

DanielHuntleySBG committed:
Trim tool icons added

Djolivald commented on readme.md:
```diff
@@ -448,7 +448,7 @@ player.source = {
     poster: '/path/to/poster.jpg',
     previewThumbnails: {
-        src: '/path/to/thumbnails.vtt'
+        src: '/path/to/thumbnails.vtt',
```
Seems like this is an error change, seems like there is always a comma even for the last element of objects

Djolivald commented on readme.md:
```diff
@@ -579,6 +579,9 @@ player.on('ready', event => {
 | `waiting`        | Sent when the requested operation (such as playback) is delayed pending the completion of another operation (such as a seek).                                                                                                                                                                                                                  |
 | `emptied`        | he media has become empty; for example, this event is sent if the media has already been loaded (or partially loaded), and the `load()` method is called to reload it.                                                                                                                                                                         |
 | `cuechange`      | Sent when a `TextTrack` has changed the currently displaying cues.                                                                                                                                                                                                                                                                             |
+| `entertrim`      | Sent when the player enters the trimming tool                                                                                                                                                                                                                                                                                                  |
```
`.` missing at the end

Djolivald commented on readme.md:
```diff
@@ -631,6 +635,10 @@ You can see the example VTT files [here](https://cdn.plyr.io/static/demo/thumbs/
 
+# Trimming
+
+It's possible to create a trim region for your video using the trim control. You will need to generate the clip yourself through listening to the 'trimchange' event and will return the start and end of the selected region.
```
```suggestion
It's possible to create a trim region from your video using the trim mode. This is useful when creating clips or repeating loops. The 'trimchange' event will give you the start and end time of the selected region.
```

Djolivald commented on readme.md:
```diff
@@ -777,7 +785,6 @@ Support this project with your organization. Your logo will show up here with a
 <a href="https://opencollective.com/plyr/organization/0/website"><img src="https://opencollective.com/plyr/organization/0/avatar.svg"></a>
 <a href="https://opencollective.com/plyr/organization/1/website"><img src="https://opencollective.com/plyr/organization/1/avatar.svg"></a><a href="https://opencollective.com/plyr/organization/2/website"><img src="https://opencollective.com/plyr/organization/2/avatar.svg"></a>
 
-
```
I would revert this

Djolivald commented on src/js/config/defaults.js:
```diff
@@ -140,6 +145,7 @@ const defaults = {
         'pip',
         'airplay',
         // 'download',
+        'trim',
```
Not sure this should be enabled by default

Djolivald commented on src/js/config/defaults.js:
```diff
@@ -365,6 +380,14 @@ const defaults = {
+            enabled: 'plyr--trim-enabled',
+            active: 'plyr--trim-active',
+            // Trim tool
+            trimTool: 'plyr__trim-tool',
```
what's the difference between the `--` and `__` ?

DanielHuntleySBG replied:
So in the rest of the code base controls use -- syntax whereas features follow the __ I agree through me off

I have also amended the other comments made, so once you are happy or if you spot anything on the other commits we can merge this in and create a PR into the main repo as well

Djolivald commented on src/js/controls.js:
```diff
@@ -53,7 +53,7 @@ const controls = {
 
             // Buttons
             this.elements.buttons = {
-                play: getElements.call(this, this.config.selectors.buttons.play),
```
is that intentional?

Djolivald commented on src/js/plugins/trim.js:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    // Get trim range in seconds
+    getTrimRange() {
```
It would probably be a good idea to have the trim range in ms aswell.
You could have `getTrimRangeSeconds` that's calling `getTrimRangeMS() / 1000`

Djolivald commented on src/js/plyr.d.ts:
```diff
@@ -450,6 +452,12 @@ declare namespace Plyr {
         captions?: CaptionOptions;
 
+        /**
+         * enabled: Toggles whether fullscreen should be enabled. fallback: Allow fallback to a full-window solution.
```
Wrong description

DanielHuntleySBG committed:
If the user is using the trimming tool, check if the seek position is outside of the trimming region and set to the start of the trimming region

theprojectsomething committed:
include `npm run serve` shortcut for `gulp serve` (useful where gulp isn't installed globally)

theprojectsomething committed:
This is a PR to allow for contextual content to be included in fullscreen (or fallback) mode. This means arbitrary elements (extensions to the basic player UI) can be overlaid and remain visible when the player switches to fullscreen.

Example use-cases include:
 - display of video title or other metadata (see the included demo)
 - alternative access to menu items, such as a searchable captions list (in cases where many hundreds of languages are available)
 - custom share dialogs
 - integrated playlists with 'playing next' overlays

This approach / PR is just an example of how this feature could work and aims to keep Plyr complexity to a minimum (while enabling some fairly interesting integrations). It utilises a single config option, and does away with the need for injecting bespoke APIs or elements into the player context on a per-project basis. Or trying to mess with what is a pretty slick, but tightly coupled system.

For the user: A new `fullscreen.container` attribute is used to provide a container selector. The container must be an ancestor of the player, otherwise it's ignored. When toggling fullscreen mode, this container is now used in place of the player. Hovering over any children of the container is the same as hovering over the controls. The exception is where the player and the child share a common ancestor (that's not the fullscreen container) ... sounds complex but it's not. You can also gain pretty fine control this way with pointer events.

Under the hood: it adds a `utils/elements/closest` helper method to find the right ancestor. If found this is returned as the fullscreen target in place of the player container. Fullscreen is instantiated slightly earlier in the setup so this container is available for the `listeners.controls` call. In here we add some more 'mouseenter/mouseleave' listeners to any direct descendants of the container, that aren't also ancestors of the player. And that's it. No extra classes, nothing else. There are some style changes to the demo (top margin on the player) but these would be project specific.

Thanks for reading.

theprojectsomething committed:
minor demo fix for fallback mode (inline style was overriding plyr stylesheet)

theprojectsomething committed:
remove dist

theprojectsomething committed:
revert demo
feature demo is available here: https://codepen.io/theprojectsomething/full/bGdyJmv

theprojectsomething committed:
convert fullscreen children listeners logic to ~ES6
remove gulp serve shortcut

DanielHuntleySBG committed:
Improved styling of Trimming timeline

DanielHuntleySBG committed:
Fix to only update the current time when the user has finished setting the trimming length (let go of trimming handle)

theprojectsomething committed:
revert dist (brain fart earlier .. deleted instead of reverting)

DanielHuntleySBG committed:
Trimming tool: Implementation of timestamp above each thumb to display the seek position in the video | Trim plugin is also brought to be more inline with the existing code base (sass structure and change of this.tool to be this.elements)

DanielHuntleySBG committed:
Improve accessiblity: Inclusion of aria metadata

DanielHuntleySBG committed:
Nit pick fixes for trimming tool PR

DanielHuntleySBG committed:
Remove plyr.svg from pull request

DanielHuntleySBG committed:
Trimming: On source change update whether or not to show the trimming tool

DanielHuntleySBG committed:
Merge branch 'video-trimming-plugin' of https://github.com/SBGSports/plyr into video-trimming-plugin

DanielHuntleySBG committed:
Video trimming bug fix: When switching sources a new trimming object was not created | We should only display the trim thumb time tooltip when displaying preview thumbnails as a seek tooltip is displayed

DanielHuntleySBG committed:
Trimming: Change structure of triggerevent messages so that they are in the correct format and pass a valid element

DanielHuntleySBG committed:
Remove Timeupdate event listener when changing source as two instances existed causing issues

DanielHuntleySBG committed:
Convert the default trim length to percent as for longer videos it was showing only a very small region

DanielHuntleySBG committed:
Addition of a couple additional api endpoints to get and set the start and end points of the trimming range

DanielHuntleySBG committed:
Update to use the existing clamp function | Improve commenting for setting the trim length | Prevent the end value of the trimming region from being before the start

sampotts committed:
Converted to 2 space indentation

DanielHuntleySBG committed:
Remove getter and setter for trimming from plyr.js as it can be accessed via the trimming object and update documentation

DanielHuntleySBG committed:
Remove settrimstart and settrimend from plyr.js as they can be set from the trim class: unsure whether a setter should exist for both of these and what its naming convention would be, so have kept it set via the variables directly

DanielHuntleySBG committed:
Calling trim.enter or trim.exit did not update the video trim control so have moved the logic to inside the enter and exit function | Additionally removed toggling trimming from plyr.js as it can all be set via trim.xxx

sampotts committed:
Merge pull request #1759 from theprojectsomething/features/fullscreen-container

Features/fullscreen container

sampotts committed:
Vimeo ratio fixes

sampotts committed:
Poster image fix (fixes #1763)

DanielHuntleySBG committed:
Fix to Trimming tool, to prevent the start time being before the end time

likev committed:
Update readme.md

There is another same Note below

likev committed:
Update readme.md

fixed code logic error

sampotts committed:
Merge pull request #1790 from likev/patch-2

Update readme.md

sampotts committed:
Merge pull request #1789 from likev/patch-1

Update readme.md

theprojectsomething committed:
This commit addresses preloading TextTracks as outlined in feature request #1791

These changes bring Plyr captions download behaviour in line with that of the default video element in major browsers. Specifically text tracks only download as they are required for display. Previously all text tracks would download when the Plyr instance was instantiated - which could become an issue when e.g. many translations are available.

For a track to be downloaded it must either be the default track, the active track when captions are toggled on, or selected from the captions menu.

sampotts committed:
Merge pull request #1793 from theprojectsomething/features/postload-captions

Preload TextTracks as per default video element

sampotts committed:
Migrate custom properties from media to parent

sampotts committed:
Use custom properties in demo

sampotts committed:
More work on custom properties and documentation

sampotts committed:
Merge branch 'develop' into css-variables

# Conflicts:
#	src/js/captions.js
#	src/js/config/defaults.js
#	src/js/fullscreen.js
#	src/js/listeners.js
#	src/js/plyr.js

sampotts committed:
Merge pull request #1770 from sampotts/css-variables

Allow customization via CSS Custom Properties

sampotts committed:
Merge

sampotts committed:
Merge branch 'develop' of github.com:sampotts/plyr into develop

# Conflicts:
#	package.json
#	yarn.lock

sampotts committed:
Merge branch 'master' into develop

# Conflicts:
#	readme.md

sampotts committed:
Fix linting issues

sampotts committed:
Housekeeping

sampotts committed:
v3.6.0

sampotts committed:
Use new feature to set thumbnails on source change in the demo

DanielHuntleySBG committed:
Merge develop branch back into video trimming branch
